util/MultiplayerCommon.h

```diff
@@ -20,6 +21,94 @@ FO_COMMON_API extern const std::string SP_SAVE_FILE_EXTENSION;
+  * "bool unused_bool = RegisterGameRules(&foo)"). */
+FO_COMMON_API bool RegisterGameRules(GameRulesFn function);
+
+/** returns the single instance of the GameRules class */
+FO_COMMON_API GameRules& GetGameRules();
+
+/** Database of values that control how the game mechanics function. */
+class FO_COMMON_API GameRules {
+public:
+    typedef OptionsDB::Option Rule;
+
+    /** \name Accessors */ //@{
+    bool    RuleExists(const std::string& name) const
+    { return m_game_rules.find(name) != m_game_rules.end(); }
+
+    /** returns the description string for rule \a rule_name, or throws
+      * std::runtime_error if no such rule exists. */
+    const std::string&  GetDescription(const std::string& rule_name) const;
+
+    /** returns the validator for rule \a rule_name, or throws
+      * std::runtime_error if no such rule exists. */
+    std::shared_ptr<const ValidatorBase> GetValidator(const std::string& rule_name) const;
+
+    /** returns all contained rules as name and value string pairs. */
+    std::vector<std::pair<std::string, std::string>> GetRulesAsStrings() const;
+
+    template <typename T>
+    T       Get(const std::string& name) const
+    {
+        auto it = m_game_rules.find(name);
+        if (it == m_game_rules.end())
+            throw std::runtime_error("GameRules::Get<>() : Attempted to get nonexistent rule \"" + name + "\".");
+        return boost::any_cast<T>(it->second.value);
+    }
+    //@}
+
+    /** \name Mutators */ //@{
+    /** adds a rule, optionally with a custom validator */
+    template <class T>
+    void    Add(const std::string& name, const std::string& description, T default_value,
+                bool engine_interal, const ValidatorBase& validator = Validator<T>())
+    {
+        auto it = m_game_rules.find(name);
+        if (it != m_game_rules.end())
+            throw std::runtime_error("GameRules::Add<>() : Rule " + name + " was added twice.");
+        m_game_rules[name] = Rule(static_cast<char>(0), name, default_value, default_value,
+                                  description, validator.Clone(), engine_interal, false, true);
+        DebugLogger() << "Added game rule named " << name << " with default value " << default_value;
+    }
+
+    template <typename T>
+    void    Set(const std::string& name, const T& value)
+    {
+        auto it = m_game_rules.find(name);
+        if (it == m_game_rules.end())
+            throw std::runtime_error("GameRules::Set<>() : Attempted to set nonexistent rule \"" + name + "\".");
+        it->second.SetFromValue(value);
+    }
+
+    void    SetFromStrings(const std::vector<std::pair<std::string, std::string>>& names_values);
+
+    /** Removes game rules that were added without being specified as
+        engine internal. */
+    void    ClearExternalRules();
+
+    /** Resets all rules to default values. */
+    void    ResetToDefaults();
+    //@}
+
+private:
+    GameRules();
+
+    std::map<std::string, Rule> m_game_rules;
```

Comment:
Use `std::unordered_map`.  

There is no reason to use `std::map`, it is slower and the primary use of this table is not to sort the rules by name.  If an alphabetically sorted list is needed for UI purposes, then do it in-place.